Add Local Incidence Card to Statistics (EXPOSUREAPP-7446)

kolyaopahle opened the pull request:
### Description
This PR Adds the local incidence card for the new local statistics feature, note that there is no logic in place to actually display the card at the moment this will be done in a follow up PR

kolyaopahle committed:
Plumbing for the local incidence cards including layouts and item view holder

NataliaLemmerth reviewed (dismissed):
UA approved

kolyaopahle committed:
Merge branch 'release/2.6.x' into feature/7446-local-incidence-card

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/statistics/ui/homecards/cards/LocalIncidenceCard.kt:
```diff
@@ -0,0 +1,93 @@
+    override val onBindData: HomeStatisticsCardsLocalIncidenceLayoutBinding.(
+        item: StatisticsCardItem,
+        payloads: List<Any>
+    ) -> Unit = { item, _ ->
```
I'm not sure if we need something like
`val curItem = payloads.filterIsInstance<Item>().singleOrNull() ?: item`
here, i.e. what the current update mechanic is in the nested recyclerview.
Will have to test more when everything is wired.

d4rken replied:
Meaning if we use the diffutil to it's fullest, as to not have the UI redraw/flash unnecessary, we could get a stale `item` when scrolling, and the newer one would be in `payloads`.

kolyaopahle replied:
seeems to work without it during my testing but i am honestly not that informed about recycler views in this context, if you feel like this is an issue let me know and ill implement it :)

d4rken replied:
When adding or removing local incidence cards, the "add local incidence" card flashes, so something with the stable IDs is not working correctly yet.
Once that is working we will have to use the payload mechanism mentioned above.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/formatter/FormatterStatistics.kt:
```diff
@@ -33,7 +33,7 @@ fun StatsItem.getPrimaryLabel(context: Context): String {
             else -> dateTimeFormatter.print(updatedAtDate)
         }
-        is IncidenceStats -> when (updatedAtDate) {
+        is IncidenceStats, is LocalIncidenceStats -> when (updatedAtDate) {
```
Do you see a way to make this nicer in a future PR?
I don't like that we keep increasing our use of "formatters", was happy to remove the old ones we had.
Maybe just maybe some kind of abstract `StatsItem`?

For another PR ofc, just  thinking out loud here :grin:

kolyaopahle replied:
Sounds good, might be hard with more dynamic data but ill keep it in mind for after the RC

d4rken commented on Corona-Warn-App/src/main/res/layout/home_statistics_cards_local_incidence_layout.xml:
```diff
@@ -0,0 +1,121 @@
+<?xml version="1.0" encoding="utf-8"?>
+<layout xmlns:android="http://schemas.android.com/apk/res/android"
```
Do we need the `<layout` here?

d4rken commented on Corona-Warn-App/src/main/res/menu/menu_statistics_local_incidence.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<menu xmlns:android="http://schemas.android.com/apk/res/android">
+    <item
+        android:id="@+id/menu_information"
```
Do these pop up in `R.id.menu` among all other IDs? Would it be useful here to make them easier to find?

kolyaopahle replied:
these id's have been used in other overflow menus before, i though this would keep it consistent 🤔

d4rken commented on Corona-Warn-App/src/main/res/layout/home_statistics_cards_local_incidence_layout.xml:
```diff
@@ -0,0 +1,121 @@
+            app:layout_constraintTop_toTopOf="@id/primary_value" />
+
+        <ImageButton
+            android:id="@+id/overflow_menu_button"
```
Do we need an accessibility label for the menu button?

kolyaopahle replied:
the contact diary card i took it from did not have any but we could probably add one

jurajkusnier committed:
Merge branch 'release/2.6.x' into feature/7446-local-incidence-card

harambasicluka committed:
Merge branch 'release/2.6.x' into feature/7446-local-incidence-card

kolyaopahle committed:
added support for addition and removal of local statistics homescreen cards,
district names are resolved in the parser,
also added fetching of local statistics based on active districts

kolyaopahle committed:
Merge remote-tracking branch 'origin/feature/7446-local-incidence-card' into feature/7446-local-incidence-card

kolyaopahle committed:
Merge branch 'release/2.6.x' into feature/7446-local-incidence-card

kolyaopahle committed:
Linting

kolyaopahle committed:
Fixed unittests

kolyaopahle committed:
Removed unused layout tag

kolyaopahle committed:
linting

jurajkusnier reviewed (changes requested):
I got crash after removing Local Incidence Card
```
    java.lang.IllegalStateException: Two different ViewHolders have the same stable ID. Stable IDs in your adapter MUST BE unique and SHOULD NOT change.
    ViewHolder 1:LocalIncidenceCard{312f8db position=2 id=8, oldPos=-1, pLpos:-1 not recyclable(1)} 
     View Holder 2:LocalIncidenceCard{ddbd4cd position=1 id=8, oldPos=-1, pLpos:-1} androidx.recyclerview.widget.RecyclerView{afd4cff VFED..... ......ID 0,269-1080,1241 #7f0a06d7 app:id/statistics_recyclerview}, adapter:de.rki.coronawarnapp.statistics.ui.homecards.StatisticsCardAdapter@9e5d0c4, layout:de.rki.coronawarnapp.statistics.ui.homecards.StatisticsLayoutManager@80dcbad, context:de.rki.coronawarnapp.ui.main.MainActivity@a39276d
```

jurajkusnier commented:
Doesn't meet the **Acceptance Criteria:**
- The user can add up to 5 additional tiles for local incidences. As long as less than 5 tiles have been added, the tile shows a button to add a new local incidence

_The code is already there but is not used. Something like this should work:_
```kotlin
// HomeFragmentViewModel.kt, line 203 
data = statsData.copy(
                            items = mutableListOf(AddStatsItem(isEnabled = statsData.items.size < 5))
                                .plus(statsData.items)
                        ),
```

jurajkusnier commented:
Doesn't meet the **Acceptance Criteria:**
- if a county was selected, a new tile for the selected county is added left to the latest added local incidence tile. The first local incidence tile is added left to the national incidence tile

kolyaopahle commented:
> I got crash after removing Local Incidence Card
> 
> ```
>     java.lang.IllegalStateException: Two different ViewHolders have the same stable ID. Stable IDs in your adapter MUST BE unique and SHOULD NOT change.
>     ViewHolder 1:LocalIncidenceCard{312f8db position=2 id=8, oldPos=-1, pLpos:-1 not recyclable(1)} 
>      View Holder 2:LocalIncidenceCard{ddbd4cd position=1 id=8, oldPos=-1, pLpos:-1} androidx.recyclerview.widget.RecyclerView{afd4cff VFED..... ......ID 0,269-1080,1241 #7f0a06d7 app:id/statistics_recyclerview}, adapter:de.rki.coronawarnapp.statistics.ui.homecards.StatisticsCardAdapter@9e5d0c4, layout:de.rki.coronawarnapp.statistics.ui.homecards.StatisticsLayoutManager@80dcbad, context:de.rki.coronawarnapp.ui.main.MainActivity@a39276d
> ```

weird, this has not happened for me during testing, will investigate

kolyaopahle committed:
Disabled local stats add card after 5 local stats are displayed

kolyaopahle committed:
Storing SelectedDistrict instead of District allows us to match the user selection to a server result and enables us to sort the cards by the time they were added at

kolyaopahle committed:
Using the district id as stable id for local incidence cards

kolyaopahle committed:
linting

harambasicluka committed:
Merge branch 'release/2.6.x' into feature/7446-local-incidence-card

kolyaopahle committed:
Hopefully this fixes the recycler view crashes

kolyaopahle committed:
Merge remote-tracking branch 'origin/feature/7446-local-incidence-card' into feature/7446-local-incidence-card

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/statistics/local/storage/LocalStatisticsConfigStorage.kt:
```diff
@@ -1,23 +1,39 @@
+    )
+
+    val activeStates = activeDistricts.flow
+        .map {
+            it.map { district -> district.district.federalStateShortName }
+                .mapNotNull { stateId -> FederalStateToPackageId.getForName(stateId) }
+                .distinct()
+        }
```
`it` should always be in the most nested scope.

Not a blocker.

jurajkusnier committed:
Merge branch 'release/2.6.x' into feature/7446-local-incidence-card

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/statistics/local/storage/LocalStatisticsConfigStorage.kt:
```diff
@@ -1,23 +1,39 @@
+    val activeStates = activeDistricts.flow
+        .map {
+            it.map { district -> district.district.federalStateShortName }
+                .mapNotNull { stateId -> FederalStateToPackageId.getForName(stateId) }
```
Maybe a little warning log if we have something that does not map?

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/statistics/local/source/LocalStatisticsProvider.kt:
```diff
@@ -41,23 +43,35 @@ class LocalStatisticsProvider @Inject constructor(
     }
 
-    val current: Flow<List<StatisticsData>> = localStatisticsData.data
+    val current: Flow<LocalStatisticsData> = localStatisticsData.data.map {
+        val groupedStats = it.reduceOrNull { acc, localStatisticsData ->
+            LocalStatisticsData(acc.items + localStatisticsData.items)
+        } ?: LocalStatisticsData()
+
+        groupedStats.copy(
+            items = groupedStats.items.filter {
+                localStatisticsConfigStorage.activeDistricts.value.any { selected ->
+                    selected.district.districtId == it.selectedDistrict?.district?.districtId
+                }
+            }.sortedBy { selected -> selected.selectedDistrict?.addedAt }.reversed()
+        )
+    }
```
Could you explain what is happening here?

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/statistics/local/source/LocalStatisticsProvider.kt:
```diff
@@ -41,23 +43,35 @@ class LocalStatisticsProvider @Inject constructor(
     }
 
-    val current: Flow<List<StatisticsData>> = localStatisticsData.data
+    val current: Flow<LocalStatisticsData> = localStatisticsData.data.map {
```
When I select  "Nordrhein-Westfallen>Aachen (Stadtregion)", nothing happens. :thinking:

d4rken replied:
Düssesldorf also  doesn't work, all items from NRW?

kolyaopahle replied:
maybe some id mismatch in the test data, will investigate

kolyaopahle replied:
the mapping for nrw to package id was indeed wrong, fixed it :)

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/statistics/local/FederalStateToPackageId.kt:
```diff
@@ -16,5 +16,10 @@ enum class FederalStateToPackageId(val packageId: Int) {
     ST(6),
     SH(4),
-    TH(6)
+    TH(6);
```
Is it intentional that some states have the same ID?

harambasicluka replied:
AFAIK one package can contain multiple states, there isn't one package per state.

kolyaopahle replied:
yep, the stats packages contain multiple states, this is just here to resolve which package maps to which states https://github.com/corona-warn-app/cwa-app-tech-spec/pull/61/files#diff-b934511166d6db28e5cc528bcfc6df5ac63dd27eb1a9cc9192a61f2b16f4b73eR140-R157

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentViewModel.kt:
```diff
@@ -187,15 +202,21 @@ class HomeFragmentViewModel @AssistedInject constructor(
                                 }
                                 else -> events.postValue(HomeFragmentEvents.GoToStatisticsExplanation)
                             }
+                        },
+                        onRemoveListener = {
+                            localStatisticsConfigStorage.activeDistricts.update { districts ->
+                                districts.filter { selected ->
+                                    selected.district.districtId != it.selectedDistrict?.district?.districtId
+                                }.toSet()
+                            }
```
Something with the removal doesn't seem to work as intended.
After playing around I end up with this (duplicates and contains already removed items):

![Screenshot from 2021-07-12 17-44-47](https://user-images.githubusercontent.com/1439229/125317070-00250800-e339-11eb-8368-3ba40cc25214.png)

kolyaopahle replied:
i've removed the nullability, maybe that already fixes it but if not i will investigate further

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/statistics/ui/homecards/cards/StatisticsCardItem.kt:
```diff
@@ -2,16 +2,24 @@ package de.rki.coronawarnapp.statistics.ui.homecards.cards
 ) : HasStableId {
 
     override val stableId: Long = when (stats) {
         is AddStatsItem -> 0L
```
Avoid `0` as stableId, "footgun waiting to happen".
Prefer Something like `AddStatsItem::class.hashCode()`

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/statistics/ui/homecards/cards/StatisticsCardItem.kt:
```diff
@@ -2,16 +2,24 @@ package de.rki.coronawarnapp.statistics.ui.homecards.cards
-        is StatsItem -> stats.cardType.id.toLong()
+        is StatsItem -> {
+            if (stats is LocalIncidenceStats) {
+                stats.selectedDistrict?.district?.districtId?.toLong() ?: 0L
```
If `AddStatsItem` has stableId `0`, and this falls back to `0L` we get a crash due to ID collision.

d4rken commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/statistics/ui/homecards/cards/StatisticsCardItem.kt:
```diff
@@ -2,16 +2,24 @@ package de.rki.coronawarnapp.statistics.ui.homecards.cards
     override val stableId: Long = when (stats) {
         is AddStatsItem -> 0L
-        is StatsItem -> stats.cardType.id.toLong()
+        is StatsItem -> {
```
I'd prefer to just split the CardItem cleanly here and have one per type, what do you think?
Neither of the listeners is valid for the `AddStatsItem` card and `onRemoveListener` is only valid for the local incidence stats.

d4rken commented on Corona-Warn-App/src/main/res/layout/home_statistics_cards_local_incidence_layout.xml:
```diff
@@ -0,0 +1,117 @@
+    android:id="@+id/incidence_container"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    tools:ignore="MissingConstraints"
```
Why are we ignoring this?
The warning usually means that in some layout/language setups, the views may jump to a specific direction due to missing constraints.

kolyaopahle replied:
no idea tbh, i stole it from home_statistics_cards_incidence_layout.xml

kolyaopahle replied:
turns out this does nothing, the ide does not complain if i remove it 😄

d4rken reviewed (changes requested):
Looking better every day :grin:.
Found a few inconsistencies we need to check out.

jurajkusnier committed:
Merge branch 'release/2.6.x' into feature/7446-local-incidence-card

harambasicluka committed:
Merge branch 'release/2.6.x' into feature/7446-local-incidence-card

jurajkusnier commented:
one minor thing: `AddCard` should be probably completely disabled when there is more than 5 `LocalIncidenceCard`. So also the ripple effect shouldn't be there.

kolyaopahle committed:
First round of addressing comments

kolyaopahle committed:
Fixed card recycling
Added distinct condition to the provided stats as the same package can contain multiple districts

kolyaopahle committed:
Merge remote-tracking branch 'origin/feature/7446-local-incidence-card' into feature/7446-local-incidence-card

harambasicluka committed:
Merge branch 'release/2.6.x' into feature/7446-local-incidence-card

harambasicluka committed:
Merge branch 'release/2.6.x' into feature/7446-local-incidence-card

kolyaopahle committed:
Refactored the Statistics Cards into different global and local subtypes

kolyaopahle committed:
Merge remote-tracking branch 'origin/feature/7446-local-incidence-card' into feature/7446-local-incidence-card

kolyaopahle committed:
Merge branch 'release/2.6.x' into feature/7446-local-incidence-card

kolyaopahle committed:
NRW is now also allowed to procure their local statistics

kolyaopahle committed:
Merge remote-tracking branch 'origin/feature/7446-local-incidence-card' into feature/7446-local-incidence-card

kolyaopahle committed:
small xml formatting changes
Disabled the AddCard completely (no ripple)

jurajkusnier reviewed (commented):
The positive trend icon should be green and the negative trend icon should be red, like the rest of the cards. But I see the trend icon only in grey color.

kolyaopahle committed:
Local Stats trend semantic is now inferred by trend

harambasicluka committed:
Merge branch 'release/2.6.x' into feature/7446-local-incidence-card

harambasicluka committed:
Merge branch 'release/2.6.x' into feature/7446-local-incidence-card
